Undo assertion on STANDALONE_WASM not working with memory growth

kripken opened the pull request:
While it isn't 100% standalone (there is an import to grow the memory), it is usable for people that implement that import, so don't break them.

fixes #9587

zeux commented:
Just a minor comment, "there is an import to grow the memory" - there actually isn't :) emscripten_resize_heap is implemented in standalone_wasm.c without any callouts to JS.

kripken commented:
Thanks @zeux! Actually that's more than a minor comment, it showed I missed something important here, and had to rework this PR.

We don't support memory growth in pure wasm runtimes yet, because of https://github.com/WebAssembly/WASI/issues/82 My goal was for this PR to make it work outside of those runtimes - that is, when standalone mode is enabled, but still running with the JS. And my assumption was that wasm runtimes would show an error on `emscripten_resize_heap` being missing, would let people know why it doesn't work (because extra non-wasi stuff is needed). But you're right, that function was already included in the wasm...

I rewrote this PR to do things properly, and added a test for growth as well:

 * Remove the assertion on memory growth not being allowed in standalone mode.
 * Add `emscripten_notify_memory_growth` which is just called to notify. That will show an error in wasm runtimes that don't support it.
 * Add a variation on libstandalone that calls that notification if growth is enabled. That lets programs not using growth still be 100% pure.
 * Add a test for memory growth in "impure" standalone wasm (i.e., standalone mode does not emit a 100% pure wasm yet).
 * Fix some bugs in `emscripten_resize_heap` and the preamble memory loading code, that the test uncovered.

VirtualTim commented:
Hey would this change also fix #9190?

kripken commented:
@VirtualTim this is separate from that, as this PR is just for standalone wasm mode - in which there is no JS. We don't have support for pthreads in that case, but on the plus side, without JS the problem in #9190 won't exist, if/when we do have standalone wasm pthreads.

kripken commented:
I also added docs for the new API.

@sbc100  maybe you want to take another look here, as it changed since the lgtm?

sbc100 commented on tools/system_libs.py:
```diff
@@ -1185,6 +1185,33 @@ class libstandalonewasm(MuslInternalLibrary):
+  def get_base_name(self):
+    name = super(libstandalonewasm, self).get_base_name()
+    if self.is_mem_grow:
+      name += '-mem_grow'
```
Seems odd to mix hyphen and underscore here.

kripken replied:
Good point, I'll make it `memgrow` (no underscore).
